### PR TITLE
fix(pin-only): judge block scalars from the whole base file

### PR DESCRIPTION
**Pull request by ivan-pinatti**
## Summary

Port of ivan-pinatti-labs/rsync-crypt#110, plus the fixes its review rounds here found.

`Pin Only` judged whether a changed workflow line sat inside a `run: |` block from three lines of diff context. When those lines held nothing shallower than the change it refused, so a pin whose step carries a comment between `name:` and `uses:` could never merge unattended; and where they did hold one, it could be wrong (a `uses:` line under an `if` inside `run: |` read as a step).

The gate job already checks out the default branch, so the script now reads the workflow file from there and judges from every line of it.

- **Trust.** The file is used only when its git blob id matches the diff's `index` line, every context and removed line matches it, and every hunk carries the line counts its header declares. The head side is rebuilt by applying the hunks.
- **No proven base, no pin.** If the base cannot be proven (main moved the file, a new file, a hunk that disagrees or is cut short), every workflow line counts as block scalar content and the diff is refused until the dependency bot rebases onto main. The three-line context judgment is removed.
- **The scan.** Openers with anchors or tags (`run: &body |`, `run: !!str |-`) are recognised; for `- name: |` and `- &step name: |` the floor is the mapping key's column; a lone indicator (`|` on its own line) marks the rest of the file as content, since its owner cannot be derived reliably and no workflow here uses one; a comment is never an opener.

## Verification

- A differential fuzz against PyYAML over roughly 160,000 valid documents finds no line the scan leaves unmarked that PyYAML reads as scalar content.
- A simulated SHA bump of every action pin in this repository's workflows grades pin-only through the whole-file path.
- The shapes CodeRabbit raised are regression tested in the repositories that carry a test suite for this gate (pre-commit-checklists, docker-torrent-box-with-vpn, rsync-crypt), and each new test fails against the version before its fix.

## Summary by CodeRabbit

* **Bug Fixes**
  * Workflow pin validation now rejects entire workflow files when their base cannot be verified against the compared revision.
  * Prevented unverified digest or version changes from being incorrectly approved.
  * Improved handling of YAML block-scalar content, anchors, tags, sequence markers, comments, indentation, and embedded script text.
  * Changes with unavailable or mismatched base content, malformed hunks, or missing diff metadata now fail safely.

* **Tests**
  * Expanded coverage for workflow pins, block-scalar edge cases, repository state mismatches, malformed diffs, and fail-closed behavior.

### CODE REVIEW

File: scripts/assert-pin-only-diff.py [hunk 1]
```diff
@@ -380,6 +411,191 @@ def _in_block_scalar(context: list[str], indent: int) -> bool:
...
+    """
+    column = _line_indent(line)
+    rest = line[column:]
+    while True:
+        marker = SEQUENCE_MARKER.match(rest)
+        if not marker:
+            return column
+        after = rest[marker.end() :]
+        if not after or after[0] in "|>&!":
+            return column
```
[the remark this answers is not in the document]
Comment (ivan-pinatti) [reply scripts/assert-pin-only-diff.py]: Confirmed with PyYAML: under `- &step name: |` (and `- !!map name: |`) the mapping starts at the property's column, and `uses:` there is a sibling key. Fixed in 8b9fe10: `_block_scalar_floor` skips leading anchors and tags and takes the dash as the floor only when the indicator itself follows (`- &body |`). Anchored and tagged compact-mapping cases added through the Pin Only path. Fail closed direction (a false refusal), applied to every port of this script.

File: scripts/assert-pin-only-diff.py [hunk 2]
```diff
@@ -457,21 +662,19 @@ def parse(diff: str) -> tuple[dict[str, tuple[Counter, Counter]], list[str]]:
...
             content = line[1:]
-            in_scalar = _in_block_scalar(old_context, _line_indent(content))
+            in_scalar = marks[0][old_number] if marks else True
             changes[path][0][normalize(content, path, in_scalar)] += 1
-            old_context.append(content)
+            old_number += 1
         elif line.startswith("+"):
             content = line[1:]
-            in_scalar = _in_block_scalar(new_context, _line_indent(content))
+            in_scalar = marks[1][new_number] if marks else True
```
[the remark this answers is not in the document]
Comment (ivan-pinatti) [reply scripts/assert-pin-only-diff.py:680]: Valid, and my previous commit message overstated it. Fixed in 1c7595f: a workflow whose base cannot be proven is now refused as a whole (a structural problem naming the unproven base), before any line is normalized, so a pip pin in a `run:` step is covered too. Added `test_refuses_a_pip_pin_in_a_workflow_with_no_base_to_read` and `..._when_main_has_moved_it` (both fail on the prior commit), plus the proven-base acceptance case. Applied to every port.
